### PR TITLE
feat: add changes to handle jina v2 base code

### PR DESCRIPTION
**Pull request by JoanFM**
PR to allow using `jinaai/jina-embeddings-v2-base-code` with `llama.cpp`. It has an extra normalization layer compared to other models of the `JinaV2` family and this is why it is considered independently.

### CODE REVIEW

**Comment by teleprint-me** [comment]
It's how the tokens are handled in llama.cpp. I'm in the middle of figuring out how `tokenizers` operates under the hood and seeing if there's a way to create a bridge between the two. Actually, your input would be invaluable (#7379). Or if you know someone that's better suited and has a deeper understanding of tokenizers (e.g. BPE/WPM) in general. I'm interested in Jina because the english version uses WPM. The spanish and dutch versions use BPE.

**Comment by JoanFM** [comment]
> It's how the tokens are handled in llama.cpp. I'm in the middle of figuring out how `tokenizers` operates under the hood and seeing if there's a way to create a bridge between the two. Actually, your input would be invaluable (#7379). Or if you know someone that's better suited and has a deeper understanding of tokenizers (e.g. BPE/WPM) in general. I'm interested in Jina because the english version uses WPM. The spanish and dutch versions use BPE. I'm more focused on Llama-2 and Llama-3 for BPE.
> 
> Aside: I have no idea how the CD/CI is setup here. I have some experience with Jenkins, but all of this is outside of the scope of what I'm focused on. Also, I'm just a contributor. I just chime in when I think I might have something of value to add.

Hey @teleprint-me ,

To be honest, I found it quite hard to work with tokenizer logic here, but I do not quite understand what you aim to achieve in #7379.  If you want we can jump in a call to discuss and make this process more agile.

**Comment by ggerganov** [comment]
> could you also guide me on how to fix the CI problems?

Rebase on latest `master` and the CI should work

**Comment by JoanFM** [comment]
> > could you also guide me on how to fix the CI problems?
> 
> Rebase on latest `master` and the CI should work

I will, thanks

**Comment by ggerganov** [comment]
> So ):\ tit should not be matched. Is there any logic in the code that eliminates these patterns \\ from the vocab?

Hm, not sure why this happens. We don't escape strings in the vocab - only in the prompt input:

https://github.com/ggerganov/llama.cpp/blob/3b38d48609280aa5f8ab7ea135a4351b2a5ee240/common/common.cpp#L249-L257

**Comment by JoanFM** [comment]
@ggerganov,

I am also trying to see if I can add support for `chinese` model and I manage to get it to work for English, but not for the Chinese characters. Is there a supported model in Chinese? So I can see if I can inspire on which tokenizers they use, etc ...?

**Comment by ggerganov** [comment]
I believe the most recent model that we added and also supports Chinese is https://huggingface.co/deepseek-ai/DeepSeek-V2. See if @fairydreaming's PR could be of any help: https://github.com/ggerganov/llama.cpp/pull/7519

**Comment by JoanFM** [comment]
Hey @ggerganov ,

I am starting to think that it is not a problem of the tokenizer. 

Here is my observation.

I am tryng to run this code to check how the embedding behaves:

```bash
gdb --args ../build/bin/embedding -m ./jina-embeddings-v2-base-code.gguf --threads 1 --verbose-prompt -p "for idx, x in enumerate(xs):\n    print(idx, x)"
```

and this is what `gdb` is telling me:

```
(gdb) run
Starting program: /home/joan/workspace/ollama/llm/llama.cpp/build/bin/embedding -m ./jina-embeddings-v2-base-code.gguf --threads 1 --verbose-prompt -p for\ idx,\ x\ in\ enumerate\(xs\):\\n\ \ \ \ print\(idx,\ x\)
```
Look at all the `\` that have been added. This seems to be the reason why I get different tokenization, in Python If I add an extra `\` before `\\n` I get the same encoding.

I am not sure if it is a problem of how the standard input is encoded or something? Do you happen to have any clue about this?

**Comment by ggerganov** [comment]
I see, does adding `-e` to the command-line argument fix the issue?

```bash
 ../build/bin/embedding -m ./jina-embeddings-v2-base-code.gguf --threads 1 --verbose-prompt -e -p "for idx, x in enumerate(xs):\n    print(idx, x)"
```

**Comment by JoanFM** [comment]
> I see, does adding `-e` to the command-line argument fix the issue?
> 
> ```shell
>  ../build/bin/embedding -m ./jina-embeddings-v2-base-code.gguf --threads 1 --verbose-prompt -e -p "for idx, x in enumerate(xs):\n    print(idx, x)"
> ```

Oh, it does!

File: convert-hf-to-gguf.py [hunk 1]
```diff
@@ -2451,11 +2448,13 @@ def __init__(self, *args, **kwargs):
 
     def get_tensors(self):
         for name, data in super().get_tensors():
-            if 'gated_layers' in name:
+            if 'gated_layer' in name:
                 d1 = data[:self.intermediate_size, :]
                 name1 = name.replace('gated_layers', 'gated_layers_w')
+                name1 = name.replace('up_gated_layer', 'gated_layers_v')
```
Comment (JoanFM) [inline convert-hf-to-gguf.py:2454]: have to fix this

File: convert-hf-to-gguf.py [hunk 2]
```diff
@@ -459,9 +456,6 @@ def get_vocab_base_pre(self, tokenizer) -> str:
         if chkhsh == "b6dc8df998e1cfbdc4eac8243701a65afe638679230920b50d6f17d81c098166":
             # ref: https://huggingface.co/allenai/OLMo-1.7-7B-hf
             res = "olmo"
-        if chkhsh == "a8594e3edff7c29c003940395316294b2c623e09894deebbc65f33f1515df79e":
```
Comment (JoanFM) [inline convert-hf-to-gguf.py:462]: have to fix this

File: convert-hf-to-gguf.py [hunk 3]
```diff
@@ -420,9 +420,6 @@ def get_vocab_base_pre(self, tokenizer) -> str:
         # NOTE: if you get an error here, you need to update the convert-hf-to-gguf-update.py script
         #       or pull the latest version of the model from Huggingface
         #       don't edit the hashes manually!
-        if chkhsh == "0ef9807a4087ebef797fc749390439009c3b9eda9ad1a097abbe738f486c01e5":
```
Comment (JoanFM) [inline convert-hf-to-gguf.py:423]: have to fix this

File: convert-hf-to-gguf.py [hunk 4]
```diff
@@ -2451,11 +2448,13 @@ def __init__(self, *args, **kwargs):
...
     def get_tensors(self):
         for name, data in super().get_tensors():
-            if 'gated_layers' in name:
+            if 'gated_layer' in name:
                 d1 = data[:self.intermediate_size, :]
                 name1 = name.replace('gated_layers', 'gated_layers_w')
+                name1 = name.replace('up_gated_layer', 'gated_layers_v')
                 d2 = data[self.intermediate_size:, :]
                 name2 = name.replace('gated_layers', 'gated_layers_v')
+                name2 = name.replace('up_gated_layer', 'gated_layers_w')
```
Comment (JoanFM) [inline convert-hf-to-gguf.py:2457]: have to fix this

**Comment by JoanFM** [comment]
> I see, does adding `-e` to the command-line argument fix the issue?
> 
> ```shell
>  ../build/bin/embedding -m ./jina-embeddings-v2-base-code.gguf --threads 1 --verbose-prompt -e -p "for idx, x in enumerate(xs):\n    print(idx, x)"
> ```

@ggerganov ,

how then can we be sure this behavior is available in the server? I see this escape option only available in the example itself.

**Comment by ggerganov** [comment]
I believe `server` already escapes these through the JSON parsing library. Btw all examples now escape by default since #7675, so no need to even add `-e` explicitly

File: llama.cpp [hunk 5]
```diff
@@ -4652,16 +4653,7 @@ static void llm_load_vocab(
 
         // for now, only BPE models have pre-tokenizers
         if (vocab.type == LLAMA_VOCAB_TYPE_BPE) {
-            if (tokenizer_pre.empty()) {
```
Comment (JoanFM) [inline llama.cpp:4655]: I will put it back

File: llama.cpp [hunk 6]
```diff
@@ -6626,7 +6623,7 @@ static int llama_model_load(const std::string & fname, llama_model & model, llam
         }
     } catch (const std::exception & err) {
         LLAMA_LOG_ERROR("%s: error loading model: %s\n", __func__, err.what());
-        return -1;
```
Comment (JoanFM) [inline llama.cpp:6629]: fix this

File: llama.cpp [hunk 7]
```diff
@@ -16243,16 +16246,23 @@ struct llama_model * llama_load_model_from_file(
         }
         model->rpc_servers.push_back(servers);
     }
-    int status = llama_model_load(path_model, *model, params);
-    GGML_ASSERT(status <= 0);
```
Comment (JoanFM) [inline llama.cpp:16247]: change back

**Comment by ggerganov** [comment]
Probably just a fluke, will restart the workflows now

**Comment by JoanFM** [comment]
@ggerganov I tested the behavior in server and works, I consider this is ready to be reviewed.

**Review by ggerganov — APPROVED** [review]

**Pull request merged**